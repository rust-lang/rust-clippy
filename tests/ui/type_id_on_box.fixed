#![warn(clippy::type_id_on_box)]

use std::any::{Any, TypeId};
use std::ops::Deref;

type SomeBox = Box<dyn Any>;

struct BadBox(Box<dyn Any>);

impl Deref for BadBox {
    type Target = Box<dyn Any>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn existential() -> impl Any {
    Box::new(1) as Box<dyn Any>
}

fn main() {
    let any_box: Box<dyn Any> = Box::new(0usize);
    let _ = (*any_box).type_id();
    let _ = TypeId::of::<Box<dyn Any>>(); // Don't lint. We explicitly say "do this instead" if this is intentional
    let _ = (*any_box).type_id();
    let any_box: &Box<dyn Any> = &(Box::new(0usize) as Box<dyn Any>);
    let _ = (**any_box).type_id(); // 2 derefs are needed here to get to the `dyn Any`

    let b = existential();
    let _ = b.type_id(); // Don't lint.

    let b: SomeBox = Box::new(0usize);
    let _ = (*b).type_id();

    let b = BadBox(Box::new(0usize));
    let _ = b.type_id(); // Don't lint. This is a call to `<BadBox as Any>::type_id`. Not `std::boxed::Box`!
}
