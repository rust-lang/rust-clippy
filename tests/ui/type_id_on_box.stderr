error: calling `.type_id()` on a `Box<dyn Any>`
  --> $DIR/type_id_on_box.rs:24:13
   |
LL |     let _ = any_box.type_id();
   |             -------^^^^^^^^^^
   |             |
   |             help: consider dereferencing first: `(*any_box)`
   |
   = note: this returns the type id of the literal type `Box<dyn Any>` instead of the type id of the boxed value, which is most likely not what you want
   = note: if this is intentional, use `TypeId::of::<Box<dyn Any>>()` instead, which makes it more clear
   = note: `-D clippy::type-id-on-box` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::type_id_on_box)]`

error: calling `.type_id()` on a `Box<dyn Any>`
  --> $DIR/type_id_on_box.rs:28:13
   |
LL |     let _ = any_box.type_id(); // 2 derefs are needed here to get to the `dyn Any`
   |             -------^^^^^^^^^^
   |             |
   |             help: consider dereferencing first: `(**any_box)`
   |
   = note: this returns the type id of the literal type `Box<dyn Any>` instead of the type id of the boxed value, which is most likely not what you want
   = note: if this is intentional, use `TypeId::of::<Box<dyn Any>>()` instead, which makes it more clear

error: calling `.type_id()` on a `Box<dyn Any>`
  --> $DIR/type_id_on_box.rs:34:13
   |
LL |     let _ = b.type_id();
   |             -^^^^^^^^^^
   |             |
   |             help: consider dereferencing first: `(*b)`
   |
   = note: this returns the type id of the literal type `Box<dyn Any>` instead of the type id of the boxed value, which is most likely not what you want
   = note: if this is intentional, use `TypeId::of::<Box<dyn Any>>()` instead, which makes it more clear

error: aborting due to 3 previous errors

